pkg/volume/volume_linux_test.go

```diff
@@ -158,5 +175,171 @@ func TestSkipPermissionChange(t *testing.T) {
+		featureGate         bool
+	}{
+		{
+			description:         "featuregate=on, fsgroupchangepolicy=always",
+			fsGroupChangePolicy: &always,
+			featureGate:         true,
+			setupFunc: func(path string) error {
+				info, err := os.Lstat(path)
+				if err != nil {
+					return err
+				}
+				// change mode of root folder to be right
+				err = os.Chmod(path, info.Mode()|expectedMask)
+				if err != nil {
+					return err
+				}
+
+				// create a subdirectory with invalid permissions
+				rogueDir := filepath.Join(path, "roguedir")
+				err = os.Mkdir(rogueDir, info.Mode())
+				if err != nil {
+					return err
+				}
+				return nil
+			},
+			assertFunc: func(path string) error {
+				rogueDir := filepath.Join(path, "roguedir")
+				hasCorrectPermissions := verifyDirectoryPermission(rogueDir, false /*readOnly*/)
+				if !hasCorrectPermissions {
+					return fmt.Errorf("invalid permissions on %s", rogueDir)
+				}
+				return nil
+			},
+		},
+		{
+			description:         "featuregate=on, fsgroupchangepolicy=onrootmismatch,rootdir=validperm",
+			fsGroupChangePolicy: &onrootMismatch,
+			featureGate:         true,
+			setupFunc: func(path string) error {
+				info, err := os.Lstat(path)
+				if err != nil {
+					return err
+				}
+				// change mode of root folder to be right
+				err = os.Chmod(path, info.Mode()|expectedMask)
+				if err != nil {
+					return err
+				}
 
+				// create a subdirectory with invalid permissions
+				rogueDir := filepath.Join(path, "roguedir")
+				err = os.Mkdir(rogueDir, rwMask)
+				if err != nil {
+					return err
+				}
+				return nil
+			},
+			assertFunc: func(path string) error {
+				rogueDir := filepath.Join(path, "roguedir")
+				hasCorrectPermissions := verifyDirectoryPermission(rogueDir, false /*readOnly*/)
+				if hasCorrectPermissions {
+					return fmt.Errorf("invalid permissions on %s", rogueDir)
+				}
+				return nil
+			},
+		},
+		{
+			description:         "featuregate=on, fsgroupchangepolicy=onrootmismatch,rootdir=invalidperm",
+			fsGroupChangePolicy: &onrootMismatch,
+			featureGate:         true,
+			setupFunc: func(path string) error {
+				// change mode of root folder to be right
+				err := os.Chmod(path, 0770)
+				if err != nil {
+					return err
+				}
+
+				// create a subdirectory with invalid permissions
+				rogueDir := filepath.Join(path, "roguedir")
+				err = os.Mkdir(rogueDir, rwMask)
+				if err != nil {
+					return err
+				}
+				return nil
+			},
+			assertFunc: func(path string) error {
+				rogueDir := filepath.Join(path, "roguedir")
+				hasCorrectPermissions := verifyDirectoryPermission(rogueDir, false /*readOnly*/)
+				if !hasCorrectPermissions {
+					return fmt.Errorf("invalid permissions on %s", rogueDir)
+				}
+				return nil
+			},
+		},
+	}
+
+	for _, test := range tests {
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ConfigurableFSGroupPolicy, test.featureGate)()
```

Comment:
Wrap it in `t.Run` so that all the local state is contained per testcase.

Reply:
fixed